src/Utility/Hash.php

```diff
@@ -336,11 +337,13 @@ public static function insert(array $data, string $path, mixed $values = null):
     /**
      * Perform a simple insert/remove operation.
      *
+     * @template T of \ArrayAccess|array
      * @param string $op The operation to do.
-     * @param \ArrayAccess|array $data The data to operate on.
+     * @param T $data The data to operate on.
      * @param list<string> $path The path to work on.
      * @param mixed $values The values to insert when doing inserts.
      * @return \ArrayAccess|array data.
+     * @psalm-return (T is array ? array : \ArrayAccess)
```

Comment:
I think this is what I was missing. I was expecting T to evaluate to a specific type in the above union.